alns/accept/GreatDeluge.py

```diff
@@ -0,0 +1,141 @@
+            raise ValueError("Alpha must be larger than 1.")
+
+        if beta <= 0 or beta >= 1:
+            raise ValueError("Beta must be in (0, 1).")
+
+        if (
+            gamma is None or delta is None or gamma <= 0 or delta <= 0
+        ) and method == "non-linear":
+            raise ValueError("Gamma and delta must be non-negative.")
+
+        if method not in ["linear", "non-linear"]:
+            raise ValueError(f"Method {method} not understrood.")
+
+        self._alpha = alpha
+        self._beta = beta
+        self._method = method
+        self._threshold = None
+
+        if self._method == "non-linear":
+            self._gamma = gamma
+            self._delta = delta
+
+    def __call__(self, rnd, best, current, candidate):
+        if self._threshold is None:
+            self._threshold = self._alpha * best.objective()
+
+            if self._method == "non-linear" and self._threshold == 0:
+                raise ValueError(
+                    "Initial solution cannot have zero objective value."
+                )
+
+        result = candidate.objective() < self._threshold
+
+        if self._method == "non-linear":
+            result |= candidate.objective() < current.objective()
+
+        self._threshold = self._update(best, current, candidate)
+
+        return result
+
+    def _update(self, best, current, candidate):
+        """
+        Return the new threshold value based on the selected GD variant.
+
+        In the linear GD variant, the threshold is updated in each iteration as
+
+        ``threshold = threshold - beta * (threshold - candidate.objective())``
+
+        TODO Improve this description.
+        In the non-linear variant, the updating method is more involved.
+        - First, the relative gap between the candidate solution and threshold
+          is computed.
+          - If this relative gap is less than `beta` parameter, meaning
+            that 1) the candidate value was higher than the threshold or 2)
+            improving less than beta relative to the threshold, then the
+            threshold is linearly increased (using the `gamma` parameter).
+          - Otherwise, the threshold is exponentially decreased (using the
+            `delta` parameter).
+        """
+
+        if self._method == "linear":
+            result = self._threshold - self._beta * (
+                self._threshold - candidate.objective()
+            )
+
+        elif self._method == "non-linear":
+            rel_gap = (
+                self._threshold - candidate.objective()
+            ) / self._threshold
+
+            if rel_gap < self._beta:
+                result = self._threshold + self._gamma * abs(
+                    candidate.objective() - self._threshold
+                )
+            else:
+                result = (
+                    self._threshold * math.exp(-self._delta * best.objective())
+                    + best.objective()
+                )
+
+        return result
```

Comment:
My first impression of this is "wow, that's complex!". I suspect we can simplify this quite a bit.